libconcord/web.cpp

```diff
@@ -106,73 +91,48 @@ static void UrlEncode(const char *in, string &out)
     }
 }
 
-static int Zap(string &server, const char *s1, const char *s2)
+#ifndef _DEBUG
+static size_t dummy_cb(char *ptr, size_t size, size_t nmemb, void *userdata)
 {
-    int err;
-
-    // Get the numeric address of host
-    hostent* addr = gethostbyname(server.c_str());
-    if (!addr) {
-        report_net_error("gethostbyname()");
-        return LC_ERROR_OS_NET;
-    }
-
-    // Fill in the sockaddr structure
-    sockaddr_in sa;
-    memcpy(&(sa.sin_addr), addr->h_addr, addr->h_length);
-    sa.sin_family = AF_INET;
-    sa.sin_port = htons(80);
-
-    // Create a socket
-    // SOCK_STREAM = Stream (TCP) socket
-    SOCKET sock = socket(sa.sin_family, SOCK_STREAM, 0);
-
-    // Connect
-    if ((err = connect(sock,(struct sockaddr*)&sa, sizeof(sa)))) {
-        report_net_error("connect()");
-        return LC_ERROR_OS_NET;
-    }
-
-    debug("Connected!");
-
-    err = send(sock, s1, strlen(s1), 0);
-    if (err == SOCKET_ERROR) {
-        report_net_error("send()");
-        return LC_ERROR_OS_NET;
-    }
-
-    debug("%i bytes sent", err);
-
-    err = send(sock, s2, strlen(s2), 0);
-    if (err == SOCKET_ERROR) {
-        report_net_error("send()");
-        return LC_ERROR_OS_NET;
-    }
+    return size * nmemb;
+}
+#endif
 
-    debug("%i bytes sent", err);
+static int Zap(string &server, string &path, string &cookie, string &post)
+{
+    int ret = 0;
+    string url = "https://" + server + "/" + path;
+
+    curl_global_init(CURL_GLOBAL_ALL);
+
+    CURL *curl = curl_easy_init();
+    if (curl) {
+        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, post.c_str());
+        curl_easy_setopt(curl, CURLOPT_USERAGENT, user_agent);
+        curl_easy_setopt(curl, CURLOPT_COOKIE, cookie.c_str());
+#ifdef _DEBUG
+        curl_easy_setopt(curl, CURLOPT_VERBOSE, 1);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, stderr);
+#else
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, dummy_cb);
+#endif
```

Comment:
what's this? If the callback is a no-op dummy anyway, why define it and pass it in?

Reply:
The default for `CURLOPT_WRITEFUNCTION` is `fwrite()`, so you actually have to define a no-op if you don't want the response written to stdout.  See: https://curl.se/libcurl/c/CURLOPT_WRITEFUNCTION.html

Reply:
Oh right. Sometimes I hate curl.